### PR TITLE
Fixing Delta Lake OOMs in warehouse sources

### PR DESCRIPTION
**Pull request by estefaniarabadan**
## Summary

Discuss how c an we solve delta OOMs

## Checklist

- [x] I have considered whether this RFC should instead be private, and confirmed it's appropriate for this public repo. (See the README — this repo is for non-product discussions that are safe to have in public. If the topic touches anything customer-sensitive, commercially sensitive, or otherwise non-public, it likely belongs in a private venue instead.)

### CODE REVIEW

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+Needs:
+
+1. **1.2 + 1.3**: these two look like the easy ones to try, even if we end up not keeping delta as the engine.
+2. **1.1**: I think we definitely need this one. The observability will help us track growing partitions, and we can filter to only track partitions bigger than X.
+3. **2.3**: No matter which route we take, separating inserts/appends/small merges from the bigger ones makes sense.
+
+We need to decide on architecture changes anyway. Since we're not sure what's going to happen with DuckDB and the managed warehouse, we should assume that if we go for DuckDB, we'll have to handle our own provisioning.
+
+**2.1 (merge-on-read)**: I really like this one; it's a nice separation of concerns. I'm just not sure how much it would affect our read layer, as I'm not familiar with it.
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:234]: It's nice in theory, but I imagine it'll be very inefficient in terms of reading the data - would be good to get a test running of how performant this really is. Plus we'd need to figure out how the managed warehouse uses this data too - currently its just copied over "as is"

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+
+---
+
+## Where my head is at
+
+Needs:
+
+1. **1.2 + 1.3**: these two look like the easy ones to try, even if we end up not keeping delta as the engine.
+2. **1.1**: I think we definitely need this one. The observability will help us track growing partitions, and we can filter to only track partitions bigger than X.
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:229]: +1

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+- ClickHouse Iceberg write support is very new (25.7–25.9), so we'd be betting on recent features and need to be on a current version; MoR also pushes cost to reads and needs a compaction cadence.
+
+
+---
+
+## Where my head is at
+
+Needs:
+
+1. **1.2 + 1.3**: these two look like the easy ones to try, even if we end up not keeping delta as the engine.
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:228]: Agreed, feels like we can add these asap

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+2. **1.1**: I think we definitely need this one. The observability will help us track growing partitions, and we can filter to only track partitions bigger than X.
+3. **2.3**: No matter which route we take, separating inserts/appends/small merges from the bigger ones makes sense.
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:230]: Anything against doing 2.2 (kubernetes jobs) - feels like thats kinda what they're made for?

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+We need to decide on architecture changes anyway. Since we're not sure what's going to happen with DuckDB and the managed warehouse, we should assume that if we go for DuckDB, we'll have to handle our own provisioning.
+
+**2.1 (merge-on-read)**: I really like this one; it's a nice separation of concerns. I'm just not sure how much it would affect our read layer, as I'm not familiar with it.
+**3.2 (Spark/EMR)**: I've used Spark a lot, so maybe I'm a bit biased here. I'm not sure how expensive EMR clusters are these days, but if we think this could be a solution, I can do a full write-up on costs, time to spin up clusters, and everything else.
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:235]: Deffo concerned with the costs here, do you have rough napkin maths of what this would look like?

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+- No single node needs to fit the partition, since Spark distributes and spills across executors.
+- Same Delta format on the same S3 path, delta-rs and ClickHouse keep reading it unchanged.
+
+**Cons**
+
+- Heavy new ops (JVM/Spark/EMR, a separate artifact, IAM).
+- Two writers on one table means pinning Delta protocol features (no deletion vectors) forever so delta-rs and ClickHouse can still read it.
+
+### 3.3 Iceberg (with ClickHouse as the merge engine)
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:204]: Would be good to test how fast MoR really is with clickhouse - whatever option we choose here, we need to make sure the managed warehouse also supports it

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+- Dynamic right-sizing can land a single huge merge no standing pool could fit, with no idle big pods.
+- Maximum isolation, each heavy merge is its own pod/node/cgroup.
+
+**Cons**
+
+- Highest operational + security lift: the app gains RBAC to create pods, plus Job lifecycle, version pinning, log routing.
+- Scheduling/cold-start latency tail (image pull + autoscaler + boot), and still has a max-node ceiling.
+
+### 2.3 Two consumer pools
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:144]: I think this is a good short term fix, but I think I'd love to have a node pool for 95% of jobs, and then some "elastic" system for the larger tables - ideally we need something that can scale infinitely (to an extent) - if we have a table with super big partitons, then ideally we can roll out a job/instance that will just handle it

File: requests-for-comments/2026-05-28-fixing-delta-oom.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+- Peak memory becomes roughly constant regardless of partition size, this actually *removes* the OOM rather than pushing the frontier out.
+- Uses shipping delta-rs primitives (replaceWhere + iterator source + schema merge); atomic, and works with our existing idempotency tagging.
+
+**Cons**
+
+- Write amplification: it rewrites the whole partition, so we only want it for big partitions (hence pair it with 1.1).
+- delta-rs replaceWhere edge cases. There are open issues like #2867 (predicate index-out-of-bounds) and #3744 (nullable property overwrite). we'd need to pin a known-good delta-rs version and cover these in tests
+
+### 1.5 Subprocess-per-merge for full isolation
```
Comment (Gilbert09) [inline requests-for-comments/2026-05-28-fixing-delta-oom.md:94]: I've tried this in the past, it's pretty tricky getting all the multiprocessing stuff to work as expected - (async, django, rust, etc etc)

**Review by Gilbert09 — APPROVED** [review]
Some good areas to explore - would be good to get some more info on costs and blockers on some of the more architectural changes - the 1.x route all seems doable and something we should start implementing sooner than later

File: requests-for-comments/2026-05-28-fixing-delta-oom.md (hunk 4 above)
Comment (estefaniarabadan) [reply requests-for-comments/2026-05-28-fixing-delta-oom.md:230]: nothing against that except for us doing the provisioning of that job (+ waiting for start time)

File: requests-for-comments/2026-05-28-fixing-delta-oom.md (hunk 5 above)
Comment (estefaniarabadan) [reply requests-for-comments/2026-05-28-fixing-delta-oom.md:235]: Not really, I can do some with the info we can get with 1.1. I can see now that they have an option to run EMR in EKS, not sure yet what this means, but if we are up to go the pyspark route, I can do a better check on price but for what claude is telling me, we can pay around $1.5 dollars per merge (I will properly check this and come back)
